tests/ranged_balance.cpp

```diff
@@ -0,0 +1,123 @@
+    shooter.wear_item( item( "backpack" ) );
+
+    // Give shooter a loaded gun of the requested type.
+    item &gun = shooter.i_add( item( gun_id ) );
+    const itype_id ammo_id = gun.ammo_default();
+    if( gun.magazine_integral() ) {
+        item &ammo = shooter.i_add( item( ammo_id, calendar::turn, gun.ammo_capacity() ) );
+	REQUIRE( gun.is_reloadable_with( ammo_id ) );
+	REQUIRE( shooter.can_reload( gun, ammo_id ) );
+	gun.reload( shooter, item_location( shooter, &ammo ), gun.ammo_capacity() );
+    } else {
+        const itype_id magazine_id = gun.magazine_default();
+        item &magazine = shooter.i_add( item( magazine_id ) );
+        item &ammo = shooter.i_add( item( ammo_id, calendar::turn, magazine.ammo_capacity() ) );
+        REQUIRE( magazine.is_reloadable_with( ammo_id ) );
+        REQUIRE( shooter.can_reload( magazine, ammo_id ) );
+        magazine.reload( shooter, item_location( shooter, &ammo ), magazine.ammo_capacity() );
+        gun.reload( shooter, item_location( shooter, &magazine ), magazine.ammo_capacity() );
+    }
+    shooter.wield( gun );
+}
+
+static statistics firing_test( npc &shooter, std::string gun_type, float aim_ratio, int range )
+{
+    equip_shooter( shooter, itype_id( gun_type ) );
+
+    // Spawn a target.
+    // Maybe change this to a random direction, or even move the target around during the test.
+    tripoint target_pos( shooter.pos() + tripoint( range, 0, 0 ) );
+    monster &test_target = spawn_test_monster( "debug_mon", target_pos );
+
+    item &gun = shooter.weapon;
+
+    dispersion_sources dispersion =
+      shooter.get_weapon_dispersion( gun,
+                                     rl_dist( shooter.pos(), test_target.pos() ) );
+    dispersion.add_range( std::max( static_cast<double>( gun.sight_dispersion() ),
+                                    shooter.recoil_total() * ( 1.0 - aim_ratio ) ) );
+
+    statistics firing_stats;
+    do {
+        // Top off target hp to make sure it doesn't die.
+        test_target.set_hp( test_target.get_hp_max() );
+        // Calling projectile_ttack() instead of fire_gun() because it returns wether it was a hit or not.
+        // Conviniently, it also avoids consuming the ammo.
+        dealt_projectile_attack shot =
+          projectile_attack( make_gun_projectile( gun ), shooter.pos(), test_target.pos(),
+                             dispersion, &shooter, nullptr );
+        firing_stats.add( shot.hit_critter != nullptr );
+    } while( firing_stats.n() < 100 ||
+             ( firing_stats.n() < 1000 && firing_stats.adj_wald_error() < 0.0001 ) );
+    return firing_stats;
+}
+
+static void test_shooting_scenario( npc &shooter, std::string gun_type,
+                                    int minimum_range, int good_range )
+{
+    clear_map();
+
+    statistics minimum_stats = firing_test( shooter, gun_type, 0.0, minimum_range );
+    INFO( "Accumulated " << minimum_stats.n() << " samples." );
+    CHECK( minimum_stats.avg() < 0.01 );
```

Comment:
I wouldn't enforce "always misses" on novices. It may be reasonable to expect it, but it only complicates things when balancing.
It does matter for experts - those certainly need to be capped from above, but again - 1% cap is really strict, at this point it's a problematic limitation that doesn't test an in-game situation.

Any hit rate below 10% that includes grazes is well into not being worth it for proper firearms. It would make sense for pneumatics or lasers, where ammunition is very cheap, but firearms are expensive to maintain and also loud.
It's basically a statistical outlier - one that is not worth including in the model.
If you bumped the rate to at least 10%, it would both become more practical and easier to actually enforce without breaking other values.

But then again, going for capping the 10% "good hit" rate would be even better. It would test actual in-game effects, not implementation details.